### DISCUSSION TITLE
BerryScript: Change serial port parity on the fly

### DISCUSSION THREAD

**Discussion by clinkme4g** [arendst/Tasmota#22272] [Support request] [answered]
I'm trying to emulate 9-bits RS-485 communication using parity bit.
The communication's rule is simple: first byte is address byte, it should be sent with first bit set (i.e. 1+8-bits byte)
Next bytes should be sent with first bit clear (0+8-bits byte).
The well known method to do this is to use 8-bits communication, but use parity bit as 9th information bit.
So, the code looks like this:
`var parity=[
        0, 1, 1, 0, 1, 0, 0, 1, 1, 0, 0, 1, 0, 1, 1, 0,
        1, 0, 0, 1, 0, 1, 1, 0, 0, 1, 1, 0, 1, 0, 0, 1,
        1, 0, 0, 1, 0, 1, 1, 0, 0, 1, 1, 0, 1, 0, 0, 1,
        0, 1, 1, 0, 1, 0, 0, 1, 1, 0, 0, 1, 0, 1, 1, 0,
        1, 0, 0, 1, 0, 1, 1, 0, 0, 1, 1, 0, 1, 0, 0, 1,
        0, 1, 1, 0, 1, 0, 0, 1, 1, 0, 0, 1, 0, 1, 1, 0,
        0, 1, 1, 0, 1, 0, 0, 1, 1, 0, 0, 1, 0, 1, 1, 0,
        1, 0, 0, 1, 0, 1, 1, 0, 0, 1, 1, 0, 1, 0, 0, 1,
        1, 0, 0, 1, 0, 1, 1, 0, 0, 1, 1, 0, 1, 0, 0, 1,
        0, 1, 1, 0, 1, 0, 0, 1, 1, 0, 0, 1, 0, 1, 1, 0,
        0, 1, 1, 0, 1, 0, 0, 1, 1, 0, 0, 1, 0, 1, 1, 0,
        1, 0, 0, 1, 0, 1, 1, 0, 0, 1, 1, 0, 1, 0, 0, 1,
        0, 1, 1, 0, 1, 0, 0, 1, 1, 0, 0, 1, 0, 1, 1, 0,
        1, 0, 0, 1, 0, 1, 1, 0, 0, 1, 1, 0, 1, 0, 0, 1,
        1, 0, 0, 1, 0, 1, 1, 0, 0, 1, 1, 0, 1, 0, 0, 1,
        0, 1, 1, 0, 1, 0, 0, 1, 1, 0, 0, 1, 0, 1, 1, 0]
        
def write(buf)
        var ch
        var ser
        ch = buf[0]
        if parity[ch] == 0
            ser = serial(-1, 39, 9600, serial.SERIAL_8O1)
        else
            ser = serial(-1, 39, 9600, serial.SERIAL_8E1)
        end
        ser.write(ch)
        ser.flush()
        ser.close()

        # or i:0..size(l)-1
        for i:1..13
            ch = buf[i]
            if parity[ch] == 1
                ser = serial(-1, 39, 9600, serial.SERIAL_8O1)
            else
                ser = serial(-1, 39, 9600, serial.SERIAL_8E1)
            end
            ser.write(ch)
            ser.flush()
            ser.close()
        end
    end
end`

The code works, but only for some calls of the output() function, then, after 20-30 calls of the output() every 1 sec,
 the esp32-s2 hangs up.
I guess that is because of full recreating the serial object very often.
Is there any methods to change the serial settings (i.e. parity) in Berry script without to recreate the entire serial object?

**Comment by s-hadinger** [comment]
You are not supposed to recreate the serial object repeatedly, it has some deep consequences for the underlying layers. If it's about changing the parity, let me explore if there are other options possible

**Reply by s-hadinger** [reply]
For clarification, the parity bit is the last one on the wire, not the first one, right? But it's the 9th bit so the most significant

**Reply by sfromis** [reply]
Indeed, serial wire order is least significant bit first, and thus the parity bit being last within the start/stop bit framing.

Would it be feasible to have two serial objects for the same pins, with different parity settings and making sure to do a flush when alternating?

**Reply by s-hadinger** [reply]
No it's not possible for two serial to be on the same gpio. But we'll find a solution

**Comment by clinkme4g** [comment]
Yes, I need change the serial port parity settings without to recreate the serial() object.
No another chance to support 9-bits communication, because ESP doesn't support it directly.
The only way is to emulate that using parity bit.

**Reply by Noschvie** [reply]
Just out of curiosity, what kind of counterpart is this?

**Comment by s-hadinger** [comment]
Good news is esp-idf does have a function `uart_set_parity()` to change parity on-the-fly. Now let's see how to connect all parts Berry-Arduino-esp-idf

**Reply by clinkme4g** [reply]
Yes, we need just integrate uart_set_parity() into Berry scripting language as serial() object method.
The device I need to communicate with is big-digits (6") seven-segments display with own RS-485 based protocol.

**Comment by s-hadinger** [comment, marked as answer]
I merged a new feature to change parity.

Now you code should be:

```berry

var parityb = bytes('9669699669969669699696699669699669969669966969969669699669969669')
var ser = serial(-1, 39, 9600, serial.SERIAL_8N1)

def write(buf)
  var ch
  var ser
  ch = buf[0]
  if b.setbits(ch, 1) == 0
    ser.config(serial.SERIAL_8O1)
  else
    ser.config(serial.SERIAL_8E1)
  end
  ser.write(ch)
  ser.flush()
end
```

I took the liberty to optimize your big array of 256 elements takes more than 4KB of RAM (which is huge). You can use a bit-array instead which consumes only 204 bytes in total.

To compute it, you can do:
```berry
var parity=[
0, 1, 1, 0, 1, 0, 0, 1, 1, 0, 0, 1, 0, 1, 1, 0,
1, 0, 0, 1, 0, 1, 1, 0, 0, 1, 1, 0, 1, 0, 0, 1,
1, 0, 0, 1, 0, 1, 1, 0, 0, 1, 1, 0, 1, 0, 0, 1,
0, 1, 1, 0, 1, 0, 0, 1, 1, 0, 0, 1, 0, 1, 1, 0,
1, 0, 0, 1, 0, 1, 1, 0, 0, 1, 1, 0, 1, 0, 0, 1,
0, 1, 1, 0, 1, 0, 0, 1, 1, 0, 0, 1, 0, 1, 1, 0,
0, 1, 1, 0, 1, 0, 0, 1, 1, 0, 0, 1, 0, 1, 1, 0,
1, 0, 0, 1, 0, 1, 1, 0, 0, 1, 1, 0, 1, 0, 0, 1,
1, 0, 0, 1, 0, 1, 1, 0, 0, 1, 1, 0, 1, 0, 0, 1,
0, 1, 1, 0, 1, 0, 0, 1, 1, 0, 0, 1, 0, 1, 1, 0,
0, 1, 1, 0, 1, 0, 0, 1, 1, 0, 0, 1, 0, 1, 1, 0,
1, 0, 0, 1, 0, 1, 1, 0, 0, 1, 1, 0, 1, 0, 0, 1,
0, 1, 1, 0, 1, 0, 0, 1, 1, 0, 0, 1, 0, 1, 1, 0,
1, 0, 0, 1, 0, 1, 1, 0, 0, 1, 1, 0, 1, 0, 0, 1,
1, 0, 0, 1, 0, 1, 1, 0, 0, 1, 1, 0, 1, 0, 0, 1,
0, 1, 1, 0, 1, 0, 0, 1, 1, 0, 0, 1, 0, 1, 1, 0]

var b = bytes()
b.resize( (size(parity)+7) / 8)
for i: 0..size(parity)-1
	b.setbits(i, 1, parity[i])
end
print(b)
# prints bytes('9669699669969669699696699669699669969669966969969669699669969669')
```
